bash_completion

```diff
@@ -2307,33 +2307,83 @@ _comp_command_offset()
                     #     *something* for every command thrown at it ($cspec != empty)
                     _minimal
                 fi
-            elif [[ $cspec == *' -F '* ]]; then
-                # complete -F <function>
+            elif [[ $cspec == *\ -[CF]\ * ]]; then
+                if [[ $cspec == *' -F '* ]]; then
+                    # complete -F <function>
 
-                # get function name
-                local func=${cspec#* -F }
-                func=${func%% *}
+                    # get function name
+                    local func=${cspec#* -F }
+                    func=${func%% *}
 
-                if ((${#COMP_WORDS[@]} >= 2)); then
-                    $func "$cmd" "${COMP_WORDS[-1]}" "${COMP_WORDS[-2]}"
+                    if ((${#COMP_WORDS[@]} >= 2)); then
+                        $func "$cmd" "${COMP_WORDS[-1]}" "${COMP_WORDS[-2]}"
+                    else
+                        $func "$cmd" "${COMP_WORDS[-1]}"
+                    fi
+
+                    # restart completion (once) if function exited with 124
+                    if (($? == 124 && retry_count++ == 0)); then
+                        # Note: When the completion function returns 124, the
+                        # state of COMPREPLY is discarded.
+                        COMPREPLY=()
+
+                        cspec=$(complete -p "$compcmd" 2>/dev/null)
+
+                        # Note: When completion spec is removed after 124, we
+                        # do not generate any completions including the default
+                        # ones. This is the behavior of the original Bash
+                        # progcomp.
+                        [[ $cspec ]] || break
+
+                        continue
+                    fi
                 else
-                    $func "$cmd" "${COMP_WORDS[-1]}"
-                fi
+                    # complete -C <command>
 
-                # restart completion (once) if function exited with 124
-                if (($? == 124 && retry_count++ == 0)); then
-                    # Note: When the completion function returns 124, the state
-                    # of COMPREPLY is discarded.
-                    COMPREPLY=()
+                    # get command name
+                    local completer=${cspec#* -C \'}
 
-                    cspec=$(complete -p "$compcmd" 2>/dev/null)
+                    # completer commands are always single-quoted
+                    if ! _comp_dequote "'$completer"; then
+                        _minimal
+                        break
+                    fi
+                    completer=${ret[0]}
+
+                    local -a suggestions
+
+                    local reset_monitor=$(shopt -po monitor) reset_lastpipe=$(shopt -p lastpipe)
```

Comment:
Do we need protection against unusual `$IFS` here (at least for eventually invoking the resets)? Cf. `git grep -F ' reset'`.

Reply:
Good point. Will fix.